### DISCUSSION TITLE
Trying to implement the algorithm but my next interval always has the same value as days elapsed

### DISCUSSION THREAD

**Discussion by syt0r** [open-spaced-repetition/free-spaced-repetition-scheduler#23] [Q&A]
I'm trying to implement v4.5 of the algorithm and have this issue. Here's [the commit](https://github.com/syt0r/Kanji-Dojo/commit/c65721433b7c5f0c8b42e034892895e071dfc8a6) with my algorithm implementation, there's also a simple unit test at the bottom

Code is in Kotlin but I've attached numbers too so if you can find anything wrong with them it would be helpful

Here's the result of execution of the `when new card is reviewed as again then interval should be less than a day` test, elapsed time in this case is 2 days, I've attached screenshot with calculations, are they valid or something seems off? 

As I understand even when elapsed time is 0 next interval should have some non 0 value, right? My algorithm returns 0 in this case

<img width="998" alt="Screenshot 2024-07-05 at 11 34 35" src="https://github.com/open-spaced-repetition/free-spaced-repetition-scheduler/assets/7478653/3e0a9d0a-2a20-490b-b30c-d895638120fb">

**Comment by L-M-Sherlock** [comment]
If the elapsed time is 0, the interval will not change.

Here is a Kotlin implementation of FSRS: https://github.com/open-spaced-repetition/android-fsrs

**Reply by syt0r** [reply]
I saw that repository, unfortunately I can't use it as is in my multiplatform project, it's not a library and it seems to be a port of typescript version and the logic there is a bit hard for me to understand, it has several weird parts too so I'm trying to make my own implementation to get to know the algorithm better in case I'll need to update or fix it. 

Hmm, I've just found one thing that makes output look promising, although I'm not sure if it's correct. In the formula I(r, S) do I need to calculate retrievability or use constant (request retention equals 0.9 or something)? I used 0.9 and now the output looks believable

```
fun testForAnswer(answer: FsrsAnswer) {
        println("Testing for answer=$answer")
        var card: FsrsCard = FsrsCard.New
        var time = now
        (0..4).forEach { dayNumber ->
            card = fsrs.getUpdatedCard(
                card = card,
                answer = answer,
                reviewTime = time
            )
            time = time + 1.days
            val interval = fsrs.getInterval(
                card = card as FsrsCard.Existing,
                now = time
            )
            println("day $dayNumber with review $answer, next interval = $interval")
        }
    }

```

<img width="575" alt="image" src="https://github.com/open-spaced-repetition/free-spaced-repetition-scheduler/assets/7478653/e8b016bd-805c-45ba-837e-34e67e2267ac">

**Reply by L-M-Sherlock** [reply]
At least, you need pass these tests: https://github.com/open-spaced-repetition/go-fsrs/blob/main/fsrs_test.go

**Reply by syt0r** [reply]
Cool, thanks, I'll check them out

**Comment by L-M-Sherlock** [comment]
Is this issue resolved?

**Reply by syt0r** [reply]
Yes, I managed to implement the scheduler too, but tests don't exactly match. Here's an example of my output:
        rust intervals: 0, 4, 15, 49, 143, 379, 0, 0, 15, 37, 85, 184, 376)
        my intervals: 0, 4, 15, 48, 140, 372, 0, 0, 14, 35, 81, 176, 361)

It's pretty similar here, but the other test where value of stability is verified has quite different result, although difficulty is correct. Looks very similar though so I'll keep my current implementation
